### DISCUSSION TITLE
Infer "env" from "dotEnv"

### DISCUSSION THREAD

**Discussion by echocrow** [vercel/turborepo#8486] [Ideas]
### Goals

1. Problem: Duplicating environment variables in env files (definition & defaults) and Turborepo configs (dependency).
2. Problem: Missing environment variable dependencies (or forgetting to remove them).
3. Goal: Set-and-forget app-specific environment variables in standard dot files, without special upkeep for Turborepo.

### Non-goals

1. Loading environment variables into the task's runtime.
2. Setting default environment variables.
3. Sharing environment variables or dotenv files.

### Background

Turborepo v2 made a push to strict mode by default. While this does help greatly with predictability and caching, it puts more burden on env variable management.

In apps where all available env variables share the same prefix, (e.g. `NEXT_` or `VITE_`), this is still manageable via one or two wildcard `"env"` entries.

However, when projects use a variety of environment variables (without a single common prefix), there is currently no OOTB method of auto-tracking these in Turborepo as dependencies.

Two example scenarios:
- SvelteKit does not enforce any common environment prefixes (besides a special `PUBLIC_` prefix for non-sensitive vars)
- Directus uses a long list of various environment variables, see e.g. their [starter .env template](https://github.com/directus/directus/blob/main/api/src/cli/utils/create-env/env-stub.liquid)

Outside of Turborepo, env variables are sometimes already tracked in a local `.env` files, sometimes including additional files such as `.env.local`, `.env.development`, etc. These provide not only defaults env values (which can still be overridden in CI/CD), but also serve  as a place to list and document relevant env variables.

At the moment, adding new env variables requires adding it to an existing  `.env` file, and then _not-forgetting-to-add_ it to `turbo.json` config, too.

This gets cumbersome and error-prone quickly, especially if `"env"` entries need to be listed multiple times in some, but not all, tasks (e.g. `gen` and `build` may depend on env vars, but not `lint`, `test` etc.).

### Proposal

Provide an opt-in feature that allows scanning env variables defined in one or more files (e.g. `.env`), and add them to the final or default list of `"env"` dependencies.

Potential config syntax (A):
```json
{
  "tasks": {
    "build": {
      "inputs": ["$TURBO_DEFAULT$", ".env", ".env.*"],
      "env": ["$TURBO_DOTENV$", "!MY_EXCEPTION_VAR"]
    }
  }
}
```
Here, `$TURBO_DOTENV$` is a special placeholder, similar to `"$TURBO_DEFAULT$`, and expands automatically at Turborepo-runtime. Whereas the latter expands into Turborepo's default input file patterns, `$TURBO_DOTENV$` might expand into all env variables defined in local `.env` files.

Challenges:
- Defining which files should/shouldn't be included in this auto-expansion.
  - Could alternatively come from a separate task property, e.g. replacing the previous `dotEnv` property.
- Parsing env files (solved problem though, assuming format is compatible with common dotenv libraries)
- ?

Alternative config syntax (B):
```json
{
  "tasks": {
    "build": {
      "dotEnv": [".env",  ".env.*"],
      "env": ["$TURBO_DEFAULT$", "!MY_EXCEPTION_VAR"]
    }
  }
}
```
Reviving the previously used `"dotEnv"` property from Turborepo v1. This property would define files to scan for environment variable definitions. Those files would be default-included in `"inputs"` (no need to list them again), and the variables defined within would be default-included in `"env"`. The `"env"` property can still be overridden, or expanded via `"$TURBO_DEFAULT$"` similar to `"inputs"`.

Challenges:
- Again parsing env files (same as above).
- ?

**Comment by anthonyshew** [comment]
Normally, you don't commit `.env` files to Git, so the `.env` wouldn't exist when you go to run your builds anywhere but on your local machine. In environments where the `.env` files are missing, what would be the expectation?

**Reply by echocrow** [reply]
@anthonyshew  Usage of `.env` file varies by frameworks and teams. We've been following Vite's default [Env Variables and Mode ](https://vitejs.dev/guide/env-and-mode) setup for some time. Not super relevant, but if you're curious, our setup looks something like this inside a given app:
- `.env`: _All_ envs used are listed here, either with sensible default configs (typically tailored for production) or empty placeholders (usually secrets with a `#SECRET` comment); git-tracked, free of secrets, and optionally includes env documentation via comments.
- `.env.development`: Small list of common local env overrides specific for development; git-tracked, free of secrets.
- `.env.local`: Env overrides, typically only used for secrets; git-ignored, and sometimes generated.
- Sometimes there may be an additional `.env.development.local` (git-ignored, e.g. for temp hot rewiring in dev) or `.env.testing`, but they are rare.

Outside of this (maybe controversial?) env composition setup, some teams commit and track a single `.env.example` file instead. Here, that file could serves as documentation for devs, and could be used to auto-config `"env"` sources for Turborepo.

**Reply by echocrow** [reply]
Going back to your question: 

> In environments where the `.env` files are missing, what would be the expectation?

In those setups, using this proposed `dotEnv` config does not make much sense.

If it was still used (perhaps erroneously, or the `.env.example` file was accidentally not committed, or somehow misspelled), it may make sense to log a warning along the lines of _"dotenv files listed in config, but none were found. no extra env variables were added"_. Alternatively, the build could also error out. For consistency, I'd follow what Turborepo already does when configuring `"inputs"` with a glob that resolves no files.

**Reply by anthonyshew** [reply]
This still doesn't sound safe since you could potentially hit cache with an outdated `.env.example` file. You're on the right track, though, that there's something Turborepo can do to be more informative about the difference between what's in your `.env` file and what is in the `env` key. I'm just not certain that these propositions are as foolproof as we'd want them to be.

**Reply by echocrow** [reply]
> This still doesn't sound safe since you could potentially hit cache with an outdated .env.example file.

Can you elaborate please?
It sounds like you're describing a scenario in which a user makes use of a new env var in src, but forgets to add it to a file  (e.g. `.env.example`) referenced in the proposed `dotEnv` turbo config. In that case, you're right: The user could get a false-positive cache hit. However, this sounds identical to the already-present issue of a user making use a new env var in src, and forgetting to add it `turbo.conf`. But I may be misunderstanding the issue you're hinting at(?).

I'm proposing this feature because, when e.g. a `.env` or `.env.example` file is present, I'd think users are more likely to remember to add it there—even if just for documentation—than to add it to the turbo config (and potentially repeatedly to different tasks).

That said, I also understand this may be a niche request. If the app does use one or two common env prefixes, those can be added to `turbo.json` instead. If a repo uses neither common env prefixes, nor a git-commited `.env`/`.env.example` file, this feature also won't help them.

**Reply by anthonyshew** [reply]
You're onto the same bit I am: If the `.env.example` or `.env` isn't right, the caching still isn't right.

I do agree that there's a better experience that we can create - but I'm not sure that automatically pulling from a `.env.example` is the way we need to go, since that has the same failure modes as not adding the variables to the `.env` key. I'd want to come up with something much more bulletproof.

I do think you're on the right track as far as reading `.env` files and providing help accordingly, though. There's something there.

**Reply by echocrow** [reply]
> I'd want to come up with something much more bulletproof.

Color me intrigued!

Off the top of my head, I'm not sure what other options there are though, besides:
- explicitly listing in `turbo.json`
  - current feature
  - bit cumbersome to re-use across tasks, and easy to forget to update
- explicitly reading from another file that already lists relevant env vars
  - initial proposal here, based on committed `.env`-like files; can grow stale
  - much like turbo config, but may be more likely to be maintained (some frameworks also won't give you types until a new env is listed in such a file first)
- implicitly infer from codebase
  - i.e. somehow magically scanning source code for actual usage; doesn't seam feasible though, due to how much env usage varies across run times, frameworks, languages, etc.

Curious if there are alternatives to help with env management and/or detection. Any ideas that come to mind?

**Reply by echocrow** [reply]
(in the meantime, we've reorganized all our env vars to start with `APP_*` or `PUBLIC_*`. this means we now have a manageable `"env"` array in `turbo.json` and won't have to remember to maintain it as we add new vars. strict mode here we come.)

**Reply by anthonyshew** [reply]
I've been thinking about if we could do something like warning when things are missing.
- When working locally, we can scan your `.env*` files and give you a warning when those variables aren't present in the `env` key.
- The addition of some sort of `turbo scan --env` that builds off of the current `turbo scan` to read your source code for instances of `process.env.*`. This could print what's missing from the `env` key (although we still may not be able to find everything, as you'll see in the last paragraph). Additionally, you could run something similar in CI and we could either hard error when things are _definitely_ missing, or warn.

My concern is that inferring env can be dangerous unless we can figure out how to get it _exactly right, every time_. The cost of over-caching is extremely high since, if it gets messed up, you can ship your preview environment variables to production. So unless there's an absolutely foolproof way of inferring what you're using, that inference is potentially dangerous.

It's also worth noting that there can be environment variables that you don't use in your source code that can drastically affect runtime behavior, like [Vitest's watch mode](https://vitest.dev/config/#watch). In this case, we can't infer anything from your source code and we would be caching the wrong thing.

**Reply by echocrow** [reply]
I like the idea of being able to validate env config. That said, I'd be mad impressed if scanning source code was viable. I don't think it is though; just to get started, we have:
- `process.env.MY_VAR`
- `const { MY_VAR } = process.env`
- `const env = z.object({...}).parse(process.env)`
- repeat above for `import.meta.env`
- framework specific env's, like SvelteKit's `$env/static/private` import module (and its dynamic & public counterparts) ([example](https://learn.svelte.dev/tutorial/env-static-private))
- envs used in bash scripts
- ~~envs from CI/CD configs, such as GitHub actions~~ (on second thought, likely not relevant for Turborepo commands?)
- env usage in other languages (python/go/rust/etc., unsure which are commonly paired with Turborepo)

Realistically, I assume support for src scanning would probably only cover a narrow slice, to the point where it may not be useful in the more complex setups where it's needed the most.

Scanning `.env` files and comparing that to a package + task config could help for some automatic checks. To be useful long term, this would need to be automated; otherwise, if one forgets to add an env to a turbo file, they'll likely also not remember to manually run the check.
But, for this to be considered viable, wouldn't that necessitate `.env` files be a valid source? At that point, why not allow inferring from them, too?

By the way, just to clarify: In this proposal, scanning `.env` files would _not_ be the default behaviour. It would be a manual opt-in config, only relevant when `.env` files are present and contain the bulk of env vars. Even then it's not intended as an all-or-nothing replacement, but as a base, that can be merged with Turbo's default envs, and additional manual envs (e.g. `CI`, `MODE`, etc.).

**Reply by anthonyshew** [reply]
The various places you'll find env vars is a good point, and not covering enough is something I've worried about with [eslint-plugin-turbo](https://www.npmjs.com/package/eslint-plugin-turbo). Thanksfully, we haven't heard feedback that it's a problem in practice (but that might be just dumb luck).

> if one forgets to add an env to a turbo file, they'll likely also not remember to manually run the check.

That was why I mentioned being able to run `turbo scan --env` (or something) in CI. You'd fail the check in that case.

> But, for this to be considered viable, wouldn't that necessitate .env files be a valid source? At that point, why not allow inferring from them, too?

My instinct is that there would be a difference in our users' expectations with these two proposals:
- Auto-infer from `.env` would effectively mean "We're going to automatically infer your environment variables." If a variable gets missed, it's easy to blame Turborepo. I would expect to hear "You said you were going to handle it?!" and, well, they'd be right. If we're going to give people the impression that they don't have to worry about it anymore, we would have to be _perfect_, and I don't think we'd be able to make it there.
- Preemptive scanning is something more aptly described as "we're going to be helpful but it's still your responsibility at the end of the day." It's certainly not as beautiful of a solution, I acknowledge, but it's also closer to what we can realistically do in our part of your stack.

(I'll also note that I like your idea generally but I'm trying to draw out the cases where it doesn't work to see if the gaps can be filled. This feature is exactly something that we've discussed amongst ourselves on core and the weaknesses have been difficult to work through and justify as something shippable.)

**Reply by echocrow** [reply]
```json
{
  "dangerouslyInferFromDotEnvAtYourOwnRiskDont@Me": [".env.example"],
}
```
/s

---

Yeah, I see your point; there's the risk users will set this without the necessary understanding. E.g. they might set this using their _git-ignored_ env file, and then wonder why their envs are all undefined in CI/CD.

I'd _like_ to think that could be improved with docs, naming, and examples, but someone's still bound to miss them. (Case in point, when we were upgrading to v2: I had already prepped a repro and almost opened an issue, wondering why `globalEnv` wasn't working—until I found [this](https://turbo.build/repo/docs/reference/package-configurations#package-configurations-can-only-override-values-in-the-tasks-key) already documented. 😅 It already has "global" in the name, and I still misinterpreted it as "applicable to all tasks (in a given package)".)

A scan feature might be less subject to misinterpretation. While that won't make configuration any easier or less verbose, it could at least help catch missing vars. (Unsure if we'd continuously use that feature ourselves, but others may have more use for it.)
